Automattic/wp-calypso
# Implement mailbox provisioning components for Email/Titan

#### Changes proposed in this Pull Request

This PR builds on the initial work in #51216, and builds out three new top-level components based on the model used by G Suite:
* `TitanMailAddMailboxes` - a page that supports account specification and serves as a replacement for the current `TitanMailQuantitySelection` component
* `TitanNewMailboxList` - a wrapper component to manage a repeating list of new mailboxes, which includes most event handling and triggering validation logic
* `TitanNewMailbox` - the lowest-level component to lay out the fields and call supplied event handlers

It is also worth noting that there is a large chunk of fully client-side validation logic currently implemented in `client/lib/titan/new-mailbox.js`, again modeled on the G Suite implementation in `client/lib/gsuite/new-users.ts` (though I am directly relying on the password verification logic implemented for G Suite).

### Known Issues
 * The back end doesn't support the alternative email address yet, so this field will be optional for now (and won't be processed in the account creation call)
 * There are still some shortcomings and limitations in the client-side validation, but I think those should be addressed in more focused PRs - a key one here is the alternative email validation
 * We are not yet validating collisions with existing mailboxes or aliases, only with the current list

#### Testing instructions

At a basic level, this is going to need some creativity to make sure the code here is robust and handles errors gracefully. To that end, the key areas we need to try and break are:

**1. Happy path testing**
This is making sure that you can do the things you'd want:
* Add one or more accounts via the UI, and try removing one or more
* Verify that once you've added those accounts, clicking on Continue correctly drops you into checkout, and if you dig into the network traffic, the shopping cart data should include the relevant fields in `extra.user_emails`.
* Verify that you can check out and that the email accounts get created. _(Note: if you're running this against a development back end, please make sure you're up to date -- the final piece this functionality was only merged earlier today.)
* Make sure that if you hit the Cancel button, that gets you out of the page.
* Make sure that if you revisit the page for a domain, you get a new set of mailboxes each time.

**2. Validation testing**
This is focused on making sure that we correctly validate inputs from the user.
* Make sure that we show errors in the UI when the user enters invalid data, e.g.: no name, no password, password too short, weird/invalid email addresses, no mailbox name -- go nuts!

**3. Flow testing**
Make sure that you can navigate to and from the page seamlessly. Ensure that you can see the following notices when it's appropriate:
* A warning when you have existing email forwards
* A notice about pricing (where the content depends on whether this is a new subscription or not)
* A notice about unused mailboxes

Also ensure that clicking on the Finish Setup link in the unused mailbox notice works.

**4. Mobile exploration**
Take the code for a spin at mobile screen sizes and make sure things look OK/reasonable.

#### Screenshots

**Before**
<img width="739" alt="titan_quantity_selection_page" src="https://user-images.githubusercontent.com/3376401/111748365-1c3bff80-8899-11eb-9f3c-8df941c4b132.png">

**After**
<img width="740" alt="titan_add_email_page" src="https://user-images.githubusercontent.com/3376401/111747729-5a84ef00-8898-11eb-9eae-280221a08370.png">

Related to #51216
